---
title: "Untitled"
author: "Alex"
date: "2024-02-13"
output: html_document
---


```{r, import arguments}
 args <- (commandArgs(trailingOnly = TRUE))
 for (i in 1:length(args)) {
   eval(parse(text = args[[i]]))
}
 
```

### Leafcutter-MD

```{r}
print(data_directory)
print(annotation_directory)
print(output_directory)
```

```{r}
library(dplyr)
library(doParallel)

get_intron_meta=function(introns){
  intron_meta=do.call(rbind,strsplit(introns,":"))
  colnames(intron_meta)=c("chr","start","end","clu")
  intron_meta=as.data.frame(intron_meta,stringsAsFactors = F)
  intron_meta$start=as.numeric(intron_meta$start)
  intron_meta$end=as.numeric(intron_meta$end)
  intron_meta$middle=.5*(intron_meta$start+intron_meta$end) # TODO: can remove this now? 
  intron_meta
}

map_clusters_to_genes=function(intron_meta, exons_table) {
    gene_df=foreach (chr=sort(unique(intron_meta$chr)), .combine=rbind) %dopar% {
    
        intron_chr=intron_meta[ intron_meta$chr==chr, ]
        exons_chr=exons_table[exons_table$chr==chr, ]

        exons_chr$temp=exons_chr$start
        intron_chr$temp=intron_chr$end
        three_prime_matches=inner_join( intron_chr, exons_chr, by="temp")

        exons_chr$temp=exons_chr$end
        intron_chr$temp=intron_chr$start
        five_prime_matches=inner_join( intron_chr, exons_chr, by="temp")

        all_matches=rbind(three_prime_matches, five_prime_matches)[ , c("clu", "gene_name")]

        all_matches=all_matches[!duplicated(all_matches),]
        
        if (nrow(all_matches)==0) return(NULL)
        all_matches$clu=paste(chr,all_matches$clu,sep=':')
        all_matches
    }

    clu_df=gene_df %>% group_by(clu) %>% summarize(genes=paste(gene_name, collapse = ","))
    
    class(clu_df)="data.frame"

    clu_df
}

```

```{r}
leafcutter_pvalues <- read.table(paste0(data_directory,"/leafcutter_outlier_pVals.txt"))
leafcutter_pvalues_ordered <- leafcutter_pvalues[,paste0("simulated_star_alignment_ERR1880",c(23:31,33:62),"Aligned.sortedByCoord.out.bam")]

gtf <- rtracklayer::import(paste0(annotation_directory,'/Homo_sapiens.GRCh38.42.primary_assembly.protein_coding.gtf'))
```


```{r message=FALSE}
exons <- read.table(paste0(annotation_directory,"/leafcutterannotation_all_exons.txt.gz"), header = T)

clusters_intron_information <- get_intron_meta(rownames(leafcutter_pvalues))
cluster_gene_information <- map_clusters_to_genes(clusters_intron_information, exons)

saveRDS(cluster_gene_information, file = paste0(output_directory, "/cluster_gene_information"))

leafcutter_pvalues_ordered$clu <- paste(clusters_intron_information$chr,clusters_intron_information$clu, sep = ":")
leafcutter_pvalues_ordered$rownames <- rownames(leafcutter_pvalues_ordered)

merged_df <- merge(cluster_gene_information,leafcutter_pvalues_ordered, by = "clu")


library(stringr)
df <- merged_df[1,]

for (i in 1:dim(cluster_gene_information)[1]){
  df <- rbind(df,data.frame("genes" = str_split(string = cluster_gene_information[i,"genes"], pattern = ",")[[1]],
                            merged_df[i,-2]))
}

df <- df[-1,]

df_genenames <- merge(df, unique(data.frame("gene_id" = gtf$gene_id, "gene_name" = gtf$gene_name)), 
                      by.x = "genes", by.y = "gene_name", all = F)

pvalues_leafcutter <- df_genenames[, c("gene_id",
                            paste0("simulated_star_alignment_ERR1880",c(23:31,33:62),"Aligned.sortedByCoord.out.bam"))]
pvalues_leafcutter_min <- as.data.frame(pvalues_leafcutter %>% group_by(gene_id) %>%
  summarise(across(paste0("simulated_star_alignment_ERR1880",c(23:31,33:62),"Aligned.sortedByCoord.out.bam"),
                   min)))

gene_names_pvalues_leafcutter <- c(unlist(pvalues_leafcutter_min[,1]))
pvalues_leafcutter_min[,1] <- NULL

rownames(pvalues_leafcutter_min) <- gene_names_pvalues_leafcutter

```

```{r}
saveRDS(pvalues_leafcutter_min, file = paste0(output_directory, "/leafcutter_gene_pvalues.RDS"))

```


```{r}
leafcutter_cluster_pvals <- read.table(paste0(data_directory,"/leafcutter_outlier_clusterPvals.txt"))
leafcutter_cluster_pvals_ordered <- leafcutter_cluster_pvals[,paste0("simulated_star_alignment_ERR1880",c(23:31,33:62),"Aligned.sortedByCoord.out.bam")]
cluster_information <- cluster_gene_information

cluster_information$clu <- gsub(x = cluster_information$clu, pattern = "\\w*:", replacement = "")


merged_df_cluster <- as.data.frame(merge(x = cluster_information, y = leafcutter_cluster_pvals_ordered, by.x = "clu", by.y = 0))


library(stringr)
df_cluster <- merged_df_cluster[1,]

for (i in 1:dim(merged_df_cluster)[1]){
  df_cluster <- rbind(df_cluster,data.frame("genes" = str_split(string = merged_df_cluster[i,"genes"], pattern = ",")[[1]],
                            merged_df_cluster[i,-2]))
}

df_cluster <- df_cluster[-1,]

df_cluster_genenames <- merge(df_cluster, unique(data.frame("gene_id" = gtf$gene_id, "gene_name" = gtf$gene_name)), 
                      by.x = "genes", by.y = "gene_name", all = F)

pvalues_leafcutter_cluster <- df_cluster_genenames[, c("gene_id",
                            paste0("simulated_star_alignment_ERR1880",c(23:31,33:62),"Aligned.sortedByCoord.out.bam"))]
pvalues_leafcutter_cluster_min <- as.data.frame(pvalues_leafcutter_cluster %>% group_by(gene_id) %>%
  summarise(across(paste0("simulated_star_alignment_ERR1880",c(23:31,33:62),"Aligned.sortedByCoord.out.bam"),
                   min)))

gene_names_pvalues_leafcutter_cluster <- c(unlist(pvalues_leafcutter_cluster_min[,1]))
pvalues_leafcutter_cluster_min[,1] <- NULL

rownames(pvalues_leafcutter_cluster_min) <- gene_names_pvalues_leafcutter_cluster


saveRDS(object = pvalues_leafcutter_cluster_min, file = paste0(output_directory, "/Leafcutter_cluster_Geuvadis_pvalues.RDS"))

```

