---
title: "Untitled"
author: "Alex"
date: "3/9/2023"
output: html_document
---


```{r}
library(FRASER)
library(data.table)
library(ggplot2)
library(precrec)
library(GenomicFeatures)
library(dplyr)
library(patchwork)
library(ggpubr)
library(cowplot)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r, import arguments}
args <- (commandArgs(trailingOnly = TRUE))
for (i in 1:length(args)) {
  eval(parse(text = args[[i]]))
}

```

```{r}
print(data_directory)
print(output_directory)
print(annotation_directory)
print(geuvadis_directory)

```




```{r}
ASpli_counts <- readRDS(file = paste0(geuvadis_directory, "/ASpli-counts.RDS"))

fRUV_bins <- readRDS(file = paste0(data_directory, "/seFit_fRUV_threshold_bins_Geuvadis.RDS"))

fRUV_junctions <- readRDS(file = paste0(data_directory, "/seFit_fRUV_threshold_junctions_Geuvadis.RDS"))

fRUV_junctions_genes <- readRDS(file = paste0(data_directory, "/seFit_fRUV_threshold_junctions_gene_Geuvadis.RDS"))



Fraser_analysis_PCA_jaccard <- loadFraserDataSet(dir = paste0(data_directory, "/Fraser"),
                            name = "Geuvadis_PCA_results_Jaccard")

Fraser_analysis_BB_jaccard <- loadFraserDataSet(dir = paste0(data_directory, "/Fraser"),
                            name = "Geuvadis_PCA_jaccard_PCA_BB_Decoder_results")

Fraser_analysis_AE_jaccard <- loadFraserDataSet(dir = paste0(data_directory, "/Fraser"),
                            name = "Geuvadis_PCA_jaccard_AE_weighted_results")



```

```{r}
outliers <- readRDS(
  file = paste0(data_directory, "/Geuvadis_outlier_genes.RDS"))
unique_outliers <- outliers[!duplicated(rownames(outliers)),]

```

```{r}
txdb <- makeTxDbFromGFF(paste0(annotation_directory,"/Homo_sapiens.GRCh38.42.primary_assembly.protein_coding.gtf"), "gtf")
```



```{r}
pvalues_fRUV_bins <- assays(fRUV_bins)$pValue
pvalues_fRUV_junctions <- assays(fRUV_junctions)$pValue
pvalues_fRUV_junctions_genes <- assays(fRUV_junctions_genes)$pValue

pvalues_Fraser_PCA_jacard <- assays(Fraser_analysis_PCA_jaccard)$pvaluesBetaBinomial_junction_jaccard
pvalues_Fraser_BB_jacard <- assays(Fraser_analysis_BB_jaccard)$pvaluesBetaBinomial_junction_jaccard
pvalues_Fraser_AE_jacard <- assays(Fraser_analysis_AE_jaccard)$pvaluesBetaBinomial_junction_jaccard


```


```{r}
df_fRUV_bins <- as.data.frame(pvalues_fRUV_bins)
df_fRUV_bins$gene <- rowData(fRUV_bins)$locus
gene_pvalues_fRUV_bins <- df_fRUV_bins %>% group_by(gene) %>% summarise(across(paste0("Sample",c(1:39)),min))
```


```{r}
fRUV_junctions_genes_annotation <- ASpli_counts@metadata$junctionCounts$gene
names(fRUV_junctions_genes_annotation) <- rownames(ASpli_counts@metadata$junctionCounts)
fRUV_junctions_genes_annotation <- fRUV_junctions_genes_annotation[names(fRUV_junctions_genes_annotation) %in% rownames(fRUV_junctions)]
order <- match(rownames(fRUV_junctions),names(fRUV_junctions_genes_annotation))
rowData(fRUV_junctions)$gene <- fRUV_junctions_genes_annotation[order]

df_fRUV_junctions <- as.data.frame(pvalues_fRUV_junctions)
df_fRUV_junctions$gene <- rowData(fRUV_junctions)$gene
gene_pvalues_fRUV_junctions <- df_fRUV_junctions %>% group_by(gene) %>% summarise(across(paste0("Sample",c(1:39)),min))

```

```{r}
df_fRUV_junctions_genes <- as.data.frame(pvalues_fRUV_junctions_genes)
df_fRUV_junctions_genes$gene <- rowData(fRUV_junctions_genes)$locus
gene_pvalues_fRUV_junctions_genes <- df_fRUV_junctions_genes %>% group_by(gene) %>% summarise(across(paste0("Sample",c(1:39)),min))
```

```{r}



query <- as.data.table(findOverlaps(rowRanges(Fraser_analysis_PCA_jaccard),genes(txdb)))

df_Fraser_PCA_jaccard <- as.data.frame(pvalues_Fraser_PCA_jacard[query$queryHits,])
df_Fraser_PCA_jaccard$gene <- genes(txdb)$gene_id[query$subjectHits]
gene_pvalues_Fraser_PCA_jaccard <- df_Fraser_PCA_jaccard %>% group_by(gene) %>%
                                                      summarise(across(paste0("Sample",c(1:39)),min))

query <- as.data.table(findOverlaps(rowRanges(Fraser_analysis_BB_jaccard),genes(txdb)))

df_Fraser_BB_jaccard <- as.data.frame(pvalues_Fraser_BB_jacard[query$queryHits,])
df_Fraser_BB_jaccard[is.na(df_Fraser_BB_jaccard)] <- 1
df_Fraser_BB_jaccard$gene <- genes(txdb)$gene_id[query$subjectHits]
gene_pvalues_Fraser_BB_jaccard <- df_Fraser_BB_jaccard %>% group_by(gene) %>%
                                                      summarise(across(paste0("Sample",c(1:39)),min))

query <- as.data.table(findOverlaps(rowRanges(Fraser_analysis_AE_jaccard),genes(txdb)))

df_Fraser_AE_jaccard <- as.data.frame(pvalues_Fraser_AE_jacard[query$queryHits,])
df_Fraser_AE_jaccard[is.na(df_Fraser_AE_jaccard)] <- 1
df_Fraser_AE_jaccard$gene <- genes(txdb)$gene_id[query$subjectHits]
gene_pvalues_Fraser_AE_jaccard <- df_Fraser_AE_jaccard %>% group_by(gene) %>%
                                                      summarise(across(paste0("Sample",c(1:39)),min))




```


```{r}
unique_outliers_fRUV_bins <- unique_outliers[rownames(unique_outliers)%in% gene_pvalues_fRUV_bins$gene,]

unique_outliers_fRUV_junctions <- unique_outliers[rownames(unique_outliers)%in% gene_pvalues_fRUV_junctions$gene,]

unique_outliers_fRUV_junctions_genes <- unique_outliers[rownames(unique_outliers)%in% gene_pvalues_fRUV_junctions_genes$gene,]

unique_outliers_Fraser_PCA_jaccard <- unique_outliers[rownames(unique_outliers)%in%
                                                             gene_pvalues_Fraser_PCA_jaccard$gene,]
unique_outliers_Fraser_BB_jaccard <- unique_outliers[rownames(unique_outliers)%in%
                                                             gene_pvalues_Fraser_BB_jaccard$gene,]
unique_outliers_Fraser_AE_jaccard <- unique_outliers[rownames(unique_outliers)%in%
                                                             gene_pvalues_Fraser_AE_jaccard$gene,]

```

```{r}
order <- match(gene_pvalues_fRUV_bins$gene, rownames(unique_outliers_fRUV_bins))
gene_pvalues_fRUV_bins <- gene_pvalues_fRUV_bins[order, -c(1)]

order <- match(gene_pvalues_fRUV_junctions$gene, rownames(unique_outliers_fRUV_junctions))
gene_pvalues_fRUV_junctions <- gene_pvalues_fRUV_junctions[order, -c(1)]

order <- match(gene_pvalues_fRUV_junctions_genes$gene, rownames(unique_outliers_fRUV_junctions_genes))
gene_pvalues_fRUV_junctions_genes <- gene_pvalues_fRUV_junctions_genes[order, -c(1)]


order <- match(gene_pvalues_Fraser_PCA_jaccard$gene, rownames(unique_outliers_Fraser_PCA_jaccard))
gene_pvalues_Fraser_PCA_jaccard <- gene_pvalues_Fraser_PCA_jaccard[order, -c(1)]

order <- match(gene_pvalues_Fraser_BB_jaccard$gene, rownames(unique_outliers_Fraser_BB_jaccard))
gene_pvalues_Fraser_BB_jaccard <- gene_pvalues_Fraser_BB_jaccard[order, -c(1)]

order <- match(gene_pvalues_Fraser_AE_jaccard$gene, rownames(unique_outliers_Fraser_AE_jaccard))
gene_pvalues_Fraser_AE_jaccard <- gene_pvalues_Fraser_AE_jaccard[order, -c(1)]


```


```{r}
intersect_names <- Reduce(list(rownames(unique_outliers_fRUV_bins),
            rownames(unique_outliers_fRUV_junctions),
            rownames(unique_outliers_fRUV_junctions_genes),
            rownames(unique_outliers_Fraser_PCA_jaccard),
            rownames(unique_outliers_Fraser_BB_jaccard),
            rownames(unique_outliers_Fraser_AE_jaccard)), f = intersect)

filter_bins <- rownames(unique_outliers_fRUV_bins) %in% intersect_names
filter_junctions <- rownames(unique_outliers_fRUV_junctions) %in% intersect_names
filter_junctions_genes <- rownames(unique_outliers_fRUV_junctions_genes) %in% intersect_names
filter_Fraser_PCA_jaccard <- rownames(unique_outliers_Fraser_PCA_jaccard) %in% intersect_names
filter_Fraser_BB_jaccard <- rownames(unique_outliers_Fraser_BB_jaccard) %in% intersect_names
filter_Fraser_AE_jaccard <- rownames(unique_outliers_Fraser_AE_jaccard) %in% intersect_names


unique_outliers_fRUV_bins <- unique_outliers_fRUV_bins[filter_bins,]
unique_outliers_fRUV_junctions <- unique_outliers_fRUV_junctions[filter_junctions,]
unique_outliers_fRUV_junctions_genes <- unique_outliers_fRUV_junctions_genes[filter_junctions_genes,]
unique_outliers_Fraser_PCA_jaccard <- unique_outliers_Fraser_PCA_jaccard[filter_Fraser_PCA_jaccard,]
unique_outliers_Fraser_BB_jaccard <- unique_outliers_Fraser_BB_jaccard[filter_Fraser_BB_jaccard,]
unique_outliers_Fraser_AE_jaccard <- unique_outliers_Fraser_AE_jaccard[filter_Fraser_AE_jaccard,]


gene_pvalues_fRUV_bins <- gene_pvalues_fRUV_bins[filter_bins,]
gene_pvalues_fRUV_junctions <- gene_pvalues_fRUV_junctions[filter_junctions,]
gene_pvalues_fRUV_junctions_genes <- gene_pvalues_fRUV_junctions_genes[filter_junctions_genes,]
gene_pvalues_Fraser_PCA_jaccard <- gene_pvalues_Fraser_PCA_jaccard[filter_Fraser_PCA_jaccard,]
gene_pvalues_Fraser_BB_jaccard <- gene_pvalues_Fraser_BB_jaccard[filter_Fraser_BB_jaccard,]
gene_pvalues_Fraser_AE_jaccard <- gene_pvalues_Fraser_AE_jaccard[filter_Fraser_AE_jaccard,]

```

```{r}

auc_fRUV_bins <- matrix(NA, ncol = ncol(gene_pvalues_fRUV_bins), nrow = 1)
    
    for(i in c(1:ncol(gene_pvalues_fRUV_bins))){
    scores <- -c(unlist(gene_pvalues_fRUV_bins[,i]))
    labels <- c(unlist(abs(unique_outliers_fRUV_bins[,i])))
    msmdat_simulation <- mmdata(c(join_scores(scores,chklen=F)),
                                c(join_labels(labels,chklen=F)))
    mscurves_simulation <- evalmod(msmdat_simulation)
    auc_fRUV_bins[,i] <- subset(auc(mscurves_simulation),curvetypes == "PRC")$aucs
      
}

auc_fRUV_junctions <- matrix(NA, ncol = ncol(gene_pvalues_fRUV_junctions), nrow = 1)
    
    for(i in c(1:ncol(gene_pvalues_fRUV_junctions))){
    scores <- -c(unlist(gene_pvalues_fRUV_junctions[,i]))
    labels <- c(unlist(abs(unique_outliers_fRUV_junctions[,i])))
    msmdat_simulation <- mmdata(c(join_scores(scores,chklen=F)),
                                c(join_labels(labels,chklen=F)))
    mscurves_simulation <- evalmod(msmdat_simulation)
    auc_fRUV_junctions[,i] <- subset(auc(mscurves_simulation),curvetypes == "PRC")$aucs
      
    }

auc_fRUV_junctions_genes <- matrix(NA, ncol = ncol(gene_pvalues_fRUV_junctions_genes), nrow = 1)
    
    for(i in c(1:ncol(gene_pvalues_fRUV_junctions_genes))){
    scores <- -c(unlist(gene_pvalues_fRUV_junctions_genes[,i]))
    labels <- c(unlist(abs(unique_outliers_fRUV_junctions_genes[,i])))
    msmdat_simulation <- mmdata(c(join_scores(scores,chklen=F)),
                                c(join_labels(labels,chklen=F)))
    mscurves_simulation <- evalmod(msmdat_simulation)
    auc_fRUV_junctions_genes[,i] <- subset(auc(mscurves_simulation),curvetypes == "PRC")$aucs
      
    }


auc_Fraser_PCA_jaccard <- matrix(NA, ncol = ncol(gene_pvalues_Fraser_PCA_jaccard), nrow = 1)
    
    for(i in c(1:ncol(gene_pvalues_Fraser_PCA_jaccard))){
    scores <- -c(unlist(gene_pvalues_Fraser_PCA_jaccard[,i]))
    labels <- c(unlist(abs(unique_outliers_Fraser_PCA_jaccard[,i])))
    msmdat_simulation <- mmdata(c(join_scores(scores,chklen=F)),
                                c(join_labels(labels,chklen=F)))
    mscurves_simulation <- evalmod(msmdat_simulation)
    auc_Fraser_PCA_jaccard[,i] <- subset(auc(mscurves_simulation),curvetypes == "PRC")$aucs
      
    }

auc_Fraser_BB_jaccard <- matrix(NA, ncol = ncol(gene_pvalues_Fraser_BB_jaccard), nrow = 1)
    
    for(i in c(1:ncol(gene_pvalues_Fraser_BB_jaccard))){
    scores <- -c(unlist(gene_pvalues_Fraser_BB_jaccard[,i]))
    labels <- c(unlist(abs(unique_outliers_Fraser_BB_jaccard[,i])))
    msmdat_simulation <- mmdata(c(join_scores(scores,chklen=F)),
                                c(join_labels(labels,chklen=F)))
    mscurves_simulation <- evalmod(msmdat_simulation)
    auc_Fraser_BB_jaccard[,i] <- subset(auc(mscurves_simulation),curvetypes == "PRC")$aucs
      
    }

auc_Fraser_AE_jaccard <- matrix(NA, ncol = ncol(gene_pvalues_Fraser_AE_jaccard), nrow = 1)
    
    for(i in c(1:ncol(gene_pvalues_Fraser_AE_jaccard))){
    scores <- -c(unlist(gene_pvalues_Fraser_AE_jaccard[,i]))
    labels <- c(unlist(abs(unique_outliers_Fraser_AE_jaccard[,i])))
    msmdat_simulation <- mmdata(c(join_scores(scores,chklen=F)),
                                c(join_labels(labels,chklen=F)))
    mscurves_simulation <- evalmod(msmdat_simulation)
    auc_Fraser_AE_jaccard[,i] <- subset(auc(mscurves_simulation),curvetypes == "PRC")$aucs
      
    }





```


```{r}
df_results <- data.frame("auc" = c(auc_fRUV_bins,
                                   auc_fRUV_junctions,
                                   auc_fRUV_junctions_genes,
                                   auc_Fraser_PCA_jaccard,
                                   auc_Fraser_BB_jaccard,
                                   auc_Fraser_AE_jaccard),
                         "method" = c(rep(c("saseR bins",
                                      "saseR ASpli",
                                      "saseR junctions",
                                      "FRASER-PCA 2.0",
                                      "FRASER-BB-Decoder 2.0",
                                      "FRASER-Autoencoder 2.0"),
                                      each = 39)))


df_results$method <- factor(df_results$method, levels = 
                              c("saseR bins",
                                      "saseR junctions",
                                      "saseR ASpli",
                                      "saseR jaccard",
                                      "FRASER-Autoencoder 2.0",
                                      "FRASER-BB-Decoder 2.0",
                                      "FRASER-PCA 2.0"))

```


```{r}
sum(unique_outliers_fRUV_bins)
sum(unique_outliers_fRUV_junctions)
sum(unique_outliers_fRUV_junctions_genes)
sum(unique_outliers_Fraser_PCA_jaccard)
sum(unique_outliers_Fraser_BB_jaccard)
sum(unique_outliers_Fraser_AE_jaccard)
```

```{r}
scores <- join_scores(-c(unlist(gene_pvalues_fRUV_bins)), 
                      -c(unlist(gene_pvalues_fRUV_junctions)),
                      -c(unlist(gene_pvalues_fRUV_junctions_genes)),
                      -c(unlist(gene_pvalues_Fraser_PCA_jaccard)),
                      -c(unlist(gene_pvalues_Fraser_BB_jaccard)),
                      -c(unlist(gene_pvalues_Fraser_AE_jaccard)),
                      chklen = F)

labels <- join_labels(c(unlist(unique_outliers_fRUV_bins)),
                      c(unlist(unique_outliers_fRUV_junctions)),
                      c(unlist(unique_outliers_fRUV_junctions_genes)),
                      c(unlist(unique_outliers_Fraser_PCA_jaccard)),
                      c(unlist(unique_outliers_Fraser_BB_jaccard)),
                      c(unlist(unique_outliers_Fraser_AE_jaccard)),
                      chklen = F)

msmdat_simulation <- mmdata(scores,
                            labels,
                            modnames =
                              c("saseR bins",
                                "saseR ASpli",
                                "saseR junctions",
                                "FRASER-PCA 2.0",
                                "FRASER-BB-Decoder 2.0",
                                "FRASER-Autoencoder 2.0"),
                            dsids = c(1:6))
mscurves_simulation <- evalmod(msmdat_simulation)
df_pr <- as.data.frame(mscurves_simulation)
df_pr$modname <- factor(df_pr$modname,levels = c("saseR bins",
                                "saseR junctions",
                                "saseR ASpli",
                                "FRASER-Autoencoder 2.0",
                                "FRASER-BB-Decoder 2.0",
                                "FRASER-PCA 2.0"
))
```



# Figure Aberrant splicing section

```{r}
bp.vals <- function(x, probs=c(0.05, 0.25, 0.5,0.75, .95)) {
  r <- quantile(x, probs=probs , na.rm=TRUE)
  names(r) <- c("ymin", "lower", "middle", "upper", "ymax")
  r
}
```

```{r}


plot_1 <- ggplot(data = df_results %>% filter(method %in% c("saseR bins",
                                      "saseR ASpli",
                                      "saseR junctions",
                                      "FRASER-PCA 2.0",
                                      "FRASER-BB-Decoder 2.0",
                                      "FRASER-Autoencoder 2.0"))
                 , aes(x = method, y = auc, col = method)) + stat_summary(fun.data = bp.vals, geom="boxplot") + geom_jitter(size = 0.2)  + 
  scale_color_manual(values = c("saseR bins" = "#E69F00", 
                                "saseR ASpli" = "#D55E00", 
                                "saseR junctions" = "#CC79A7",
                                "FRASER-PCA 2.0"= "#009E73", 
                                "FRASER-BB-Decoder 2.0" = "#56B4E9",
                                "FRASER-Autoencoder 2.0" = "#0072B2")) + 
  ylab("Area under the curve per sample") + theme_bw()  + ylim(c(0,1.0001)) + theme(legend.position = "none")+
  theme(axis.text=element_text(size=10), axis.title=element_text(size=12)) + theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x =)
print(plot_1)
```



```{r}
plot_2 <- ggplot(data = df_pr%>% filter(type == "PRC" & (modname %in% c("saseR bins",
                                "saseR ASpli",
                                "saseR junctions",
                                "FRASER-Autoencoder 2.0",
                                      "FRASER-BB-Decoder 2.0",
                                      "FRASER-PCA 2.0"))), aes(x = x, y = y, col = modname)) + geom_line() + ggtitle(label = "") + ylab("Precision")  + xlab("Recall") + theme_bw() + 
  theme(axis.text=element_text(size=10), axis.title=element_text(size=12))  + scale_color_manual(values = 
                              c("saseR bins" = "#E69F00", 
                                "saseR ASpli"="#D55E00", 
                                "saseR junctions" = "#CC79A7",
                                "FRASER-PCA 2.0"= "#009E73", 
                                "FRASER-BB-Decoder 2.0" = "#56B4E9",
                                "FRASER-Autoencoder 2.0" = "#0072B2"), 
                              breaks = c("saseR bins", 
                                "FRASER-Autoencoder 2.0", 
                                "saseR junctions", 
                                "FRASER-BB-Decoder 2.0",
                                "saseR ASpli",
                                "FRASER-PCA 2.0"))+ 
  theme( legend.title = element_blank(),legend.position = "bottom", 
         legend.text = element_text(size = 12))
print(plot_2)

```

```{r}
legend <- get_legend(plot_2)
prow <- plot_grid(plot_1 + theme(legend.position = "none"), plot_2 + theme(legend.position = "none"),
                  labels = c("A","B"), ncol = 2,
                  rel_heights = c(1,1), align = "h")
Figure_1 <- plot_grid(prow,legend,ncol = 1, rel_heights= c(1,0.1))
Figure_1

```


```{r}
ggsave(paste0(output_directory,"/Supplementary-Figure-9.png"), Figure_1, 
       device = "png", width = 9, height = 4.5, dpi = 300)
ggsave(paste0(output_directory,"/Supplementary-Figure-9.eps"), Figure_1, 
       device = cairo_ps, width = 9, height = 4.5, dpi = 300)
ggsave(paste0(output_directory,"/Supplementary-Figure-9.pdf"), Figure_1, 
       device = "pdf", width = 9, height = 4.5, dpi = 300)

```
